workers/unity/Packages/com.improbable.gdk.mobile/Editor/LaunchMenu.cs

```diff
@@ -70,25 +75,94 @@ private static void LaunchMobileClient()
                         "-e", "\"arguments\"", $"\\\"{arguments.ToString()}\\\"")
                     .Run();
-
-                EditorUtility.DisplayProgressBar("Launching Mobile Client", "Done", 1.0f);
             }
             finally
             {
                 EditorUtility.ClearProgressBar();
             }
         }
 
-        private static bool TryGetApkPath(string rootPath, out string apkPath)
+        [MenuItem(MenuLaunchiOSDevice, false, 74)]
+        private static void LaunchiOSDeviceClient()
         {
-            foreach (var file in Directory.GetFiles(rootPath, "*.apk", SearchOption.AllDirectories))
+            try
             {
-                apkPath = file;
-                return true;
-            }
+                // Ensure needed tools are installed
+                if (string.IsNullOrEmpty(LibIDeviceInstallerBinary))
+                {
+                    Debug.LogError("Could not find ideviceinstaller tool. Please ensure it is installed. " +
+                        "See https://github.com/libimobiledevice/ideviceinstaller fore more details.");
+                    return;
+                }
+
+                if (string.IsNullOrEmpty(LibIDeviceDebugBinary))
+                {
+                    Debug.LogError("Could not find idevicedebug tool. Please ensure libimobiledevice is installed. " +
+                        "See https://helpmanual.io/help/idevicedebug/ for more details.");
+                    return;
+                }
+
+                EditorUtility.DisplayProgressBar("Launching iOS Device Client", "Installing archive", 0.3f);
+
+                // Get chosen ios package id
+                var bundleId = PlayerSettings.GetApplicationIdentifier(BuildTargetGroup.iOS);
+                string appList = null;
+
+                // Find archive to install
+                var ipaPath = Directory.GetFiles(AbsoluteAppBuildPath, "*.ipa", SearchOption.AllDirectories).FirstOrDefault();
+                RedirectedProcess.Command(LibIDeviceInstallerBinary).WithArgs("-l")
+                    .AddAccumulatedOutputProcessing(output => { appList = output; })
+                    .RedirectOutputOptions(OutputRedirectBehaviour.None).Run();
+                var existsOnDevice = appList.Contains(bundleId);
+                var existsLocally = !string.IsNullOrEmpty(ipaPath);
+                
+                // If .ipa exists - we install/upgrade device app and launch it
+                // If .ipa doesn't exist, but the bunlde is installed on device - we launch existing app with parameters
+                // If .ipa doesn't exist and bundle isn't found on device - we fail with an error.
+                if (!existsLocally && !existsOnDevice)
+                {
+                    Debug.LogError($"Could not find an app on device or built out iOS .ipa archive in \"{AbsoluteAppBuildPath}\" to launch.");
+                    return;
+                }
+
+                if (existsLocally)
```

Comment:
Doesn't this imply that we reinstall the app every single time, even if the operation is not required? 

Not sure how long this normally takes, but if it takes a while - it may be worth looking to check if its a new IPA or the same one that's installed.

Reply:
Not exactly, in a line below, where we actually do install, we use `-g` option if the archive both exists on device and locally, which triggers app "upgrade" instead of direct reinstall.
Let me add this to the comment.

Reply:
does this app upgrade always work? I know that for Android, if an app has been built from different machines / different signatures, the upgrade will fail

Reply:
Dug into the code a bit and it looks like `update` and `install` commands are being delivered to Apple system services, where it is hard to understand what is the logic applied.
More general, .ipa installations happen from time to time on iOS, when the app with the same bundle id has different entitlements (access to system services) or in some scenarios with codesigning differences. Though in my experience, those are more likely to happen when we install directly, not try to upgrade the app and the common workaround is to just remove existing app.
Considering that, I would imagine using `upgrade` if the bundle is already present won't make our workflows more fragile, but will save a bit of time every now and then